impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -546,8 +558,11 @@ public boolean hasNext() {
         
         final Map.Entry<K, OnHeapValueHolder<V>> thisEntry = next;
         advance();
-    
-        setAccessTimeAndExpiry(thisEntry.getKey(), thisEntry.getValue(), timeSource.getTimeMillis());
+
+        if (setAccessTimeAndExpiry(thisEntry.getKey(), thisEntry.getValue(), timeSource.getTimeMillis()) == null) {
+          onExpiration(thisEntry.getKey(), thisEntry.getValue());
```

Comment:
`onExpiration` cannot be called while not under lock, and nothing here will remove the mapping. Find another way of solving this.